src/methods/graphs.jl

```diff
@@ -51,7 +51,70 @@ function disconnected_components(g::TrophicGraph)
                 add_edge!(sub, sub_source, sub_target)
             end
         end
-        TrophicGraph(sub, labels, revmap, _original_index)
+        TrophicGraph(sub, labels, revmap, _original_species)
     end
 end
 export disconnected_components
+
+"""
+    isolated_producers(g::TrophicGraph)
+
+Collect isolated producers nodes in the trophic graph
+*i.e.* producer without incoming edges.
+"""
+function isolated_producers(graph::TrophicGraph)
+    (; _revmap, _original_species, _topology) = graph
+    res = Set{Symbol}()
+    for (i, sp) in enumerate(_revmap)
+        is_producer = _original_species[sp]
+        is_producer || continue
+        isempty(inneighbors(_topology, i)) || continue
+        push!(res, sp)
+    end
+    res
+end
+export isolated_producers
+
+"""
+    starving_consumers(graph::TrophicGraph)
+
+Collect starving consumers nodes in the trophic graph
+*i.e.* consumers with no directed path to a consumer.
+"""
+function starving_consumers(graph::TrophicGraph)
+    (; _topology, _original_species, _revmap) = graph
+
+    # Collect all current producers and consumers.
+    producers = Vector{Int}()
+    consumers = Set{Symbol}()
+    for (i, sp) in enumerate(_revmap)
+        is_producer = _original_species[sp]
+        if is_producer
+            push!(producers, i)
+        else
+            push!(consumers, sp)
+        end
+    end
+
+    # Visit the graph from producers up to consumers,
+    # and remove all consumers founds.
+    to_visit = producers
+    found = Set{Int}()
+    while !isempty(to_visit)
+        i = pop!(to_visit)
+        sp = _revmap[i]
+        is_consumer = !_original_species[sp]
+        if is_consumer
+            pop!(consumers, sp)
+        end
+        push!(found, i)
+        for up in inneighbors(_topology, i)
+            up in found && continue
+            push!(to_visit, up)
+        end
+    end
+
+    # The remaining consumers are starving.
+    consumers
+end
+export starving_consumers
```

Comment:
Wow it looks so great @iago-lito!
I have myself worked on a workaround, much simpler than yours but essentially does the same thing I think,
i.e. idenfiying starving consumers and isolated species.

```julia
"""
         connected_graph(A, producer_indexes)

removed consumers not connected to a producer (starving consumers) and isolated producers.

# Argument

- A: Adjacency matrix
- producer_indexes: position of producers in the adjacency matrix

# Value

return a tuple:
- graph_component: the graph components connected to a producer
- species: all the species connected

# Examples

# Consumer 3 is connected to consumer 2 which has no more feeding links
A = [0 0 0; 0 0 0; 0 1 0]
producers_idxs = [1]
expected = Int64[]
expected == connected_graph(A, producers_idxs).species

# Consumer 3 is connected to consumer 2 which is connected to a producer
A = [0 0 0; 1 0 0; 0 1 0]
producers_idxs = [1]
expected = Int64[1, 2, 3]
expected == connected_graph(A, producers_idxs).species

# Producer 1 is disconnected
A = [0 0 0; 0 0 0; 0 1 0]
producers_idxs = [1, 2]
expected = [2, 3]
expected == connected_graph(A, producers_idxs).species

# Only producers and all disconnected
A = [0 0 0; 0 0 0; 0 0 0]
producers_idxs = [1, 2, 3]
expected = Int64[]
expected == connected_graph(A, producers_idxs).species

# Only disconnected consumers
A = [0 0; 0 0]
producers_idxs = []
expected = Int64[]
expected == connected_graph(A, producers_idxs).species

# Two trophic chains and two producers
A = [0 0 0 0; 0 0 0 0; 1 0 0 0; 0 1 0 0]
producers_idxs = [1, 2]
expected = [1, 2, 3, 4]
all(expected .== sort(connected_graph(A, producers_idxs).species)) == true

# Two trophic chains on one producer
A = [0 0 0 0; 1 0 0 0; 0 1 0 0; 1 0 0 0]
producers_idxs = [1]
expected = [1, 2, 3, 4]
all(expected .== sort(connected_graph(A, producers_idxs).species)) == true

# Two trophic chains on one isolated producer
A = [0 0 0 0; 0 0 0 0; 1 0 0 0; 1 0 0 0]
producers_idxs = [1, 2]
expected = [1, 3, 4]
all(expected .== sort(connected_graph(A, producers_idxs).species)) == true

# One trophic chain and one isolated consumer (4)
A = [0 0 0 0; 1 0 0 0; 0 1 0 0; 0 0 0 0]
producers_idxs = [1]
expected = [1, 2, 3]
all(expected .== sort(connected_graph(A, producers_idxs).species)) == true

"""
function connected_graph(A, producer_indexes)
    # Get all the sepated components of the graph
    graph_component = connected_components(SimpleDiGraph(A))
    # Get the components that are connected to a producer
    connected_to_producers = [
                              length(intersect(producer_indexes, graph_component[i])) > 0
                              for i in 1:length(graph_component)
                             ]
    # Get the isolated components (i.e. number of species == 1)
    isolated_component = [length(graph_component[i]) == 1 for i in 1:length(graph_component)]
    # filter
    to_keep = connected_to_producers .& (.!isolated_component)

    # get the component to keep and the corresponding species
    graph_to_keep = graph_component[to_keep]
    species_idxs_to_keep = reduce(vcat, graph_to_keep, init = Int64[])
    (graph_components = graph_to_keep, species = species_idxs_to_keep,)
end
```

Reply:
Wops! Sorry that I have been so long that you had to implement this workaround by yourself @alaindanet :(
I have just finished the PR now if you'd like to review :)